ccmlib/cluster.py

```diff
@@ -106,36 +106,78 @@ def set_install_dir(self, install_dir=None, version=None, verbose=False):
+            This class is embedded here for now, because it is used only from
+            within Cluster, and depends on cluster.nodelist().
+            """
+            def __init__(self, cluster):
+                super(LogWatchingThread, self).__init__()
+                self.cluster = cluster
+                self.daemon = True  # set so that thread will exit when main thread exits
+                self.req_stop_event = threading.Event()
+                self.done_event = threading.Event()
+                self.log_positions = defaultdict(int)
+
+            def scan(self):
+                errordata = OrderedDict()
 
-        def grep_and_mark_all_logs():
-            errordata = OrderedDict()
+                try:
+                    for node in self.cluster.nodelist():
+                        errors = node.grep_log_for_errors_from(seek_start=self.log_positions[node.name])
+                        self.log_positions[node.name] = node.mark_log()
+                        if errors:
+                            errordata[node.name] = errors
+                except IOError as e:
+                    if 'No such file or directory' in e.strerror:
+                        pass  # most likely log file isn't yet written
+
+                    # in the case of unexpected error, report this thread to the callback
+                    else:
+                        errordata['log_scanner'] = [[str(e)]]
 
-            try:
-                for node in self.nodelist():
-                    errors = node.grep_log_for_errors_from(seek_start=log_positions[node.name])
-                    log_positions[node.name] = node.mark_log()
-                    if errors:
-                        errordata[node.name] = errors
-            except IOError as e:
-                if 'No such file or directory' in e.strerror:
-                    pass  # most likely log file isn't yet written
-                else:
-                    raise
-
-            if errordata:
-                on_error_call(errordata)
-
-            timer = threading.Timer(interval, grep_and_mark_all_logs)
-            timer.daemon = True  # so it will exit when the main process exits
-            timer.start()
-
-        grep_and_mark_all_logs()
+                return errordata
+
+            def run(self):
+                print_("Log-watching thread starting.")
+
+                # run until stop gets requested by .join()
+                while not self.req_stop_event.is_set():
+                    errordata = self.scan()
+
+                    if errordata:
+                        on_error_call(errordata)
+
+                    time.sleep(interval)
+
+                try:
+                    # do a final scan to make sure we got to the very end of the files
+                    errordata = self.scan()
+                    if errordata:
+                        on_error_call(errordata)
+                finally:
+                    print_("Log-watching thread exiting.")
```

Comment:
This is super not your fault, since I let this sit for 6 years before reviewing, but can you use the new logging code inside `common.py` instead of printing?